File: qml/Shimmer.js
.pragma library

// Port of shinyShimmerStep / shinyPinnedHeading / shinyEffectTickMs from
// hypr-shiny-border runtime.cpp (cursor/gradient-steps-bd3a). The walk
// stays on the CPU: two independent channels, each smoothstep-easing
// toward a random target, then picking a new target and duration
// (0.6–1.4 of 1/hz). Do not put this in GLSL.

var TAU = Math.PI * 2
var XORSHIFT_FALLBACK = 0x9E3779B9

function wrapAngle(radians) {
  var r = radians % TAU
  if (r < 0)
    r += TAU
  return r
}

// pinDeg + offsetDeg, degrees CCW → radians in [0, 2π).
// atan(-y, x): 0 = right, 90 = up.
function pinnedHeading(pinDeg, offsetDeg) {
  return wrapAngle((Number(pinDeg) + Number(offsetDeg)) * Math.PI / 180)
}

// ~32 samples per 1/hz cycle, clamped 16–50 ms. Same as shinyPulseTickMs.
// Shader twin: pulse off (brightness <= 0) is identity 1. Pulse on uses
// the same 0.5+0.5*sin that used to breathe spread/thick, now as a
// multiplier on sampled stop alpha. Twin of shinyPulseAlphaMul / GLSL.
function pulseAlphaMul(brightness, time) {
  if (!(Number(brightness) > 0))
    return 1
  return 0.5 + 0.5 * Math.sin(Number(time) * Number(brightness) * TAU)
}

function tickMs(hz) {
  var kMin = 16
  var kMax = 50
  if (!(hz > 0))
    return kMin
  var sampled = Math.floor((1000 / hz) / 32)
  if (sampled < kMin)
    return kMin
  if (sampled > kMax)
    return kMax
  return sampled
}

function makeChannel(value) {
  return { value: value, from: value, to: value, t: 0, dur: 0 }
}

function makeState(seed) {
  var s = {
    angle: makeChannel(0),
    scale: makeChannel(1),
    rng: XORSHIFT_FALLBACK
  }
  seedState(s, seed)
  return s
}

function seedState(s, seed) {
  var n = seed >>> 0
  s.rng = n !== 0 ? n : XORSHIFT_FALLBACK
}

function xorshift32(s) {
  var x = s.rng >>> 0
  x = (x ^ ((x << 13) >>> 0)) >>> 0
  x = (x ^ (x >>> 17)) >>> 0
  x = (x ^ ((x << 5) >>> 0)) >>> 0
  s.rng = x
  return x
}

function rand01(s) {
  return xorshift32(s) / 4294967296
}

function retarget(channel, s, lo, hi, hz) {
  channel.from = channel.value
  channel.to = lo + (hi - lo) * rand01(s)
  channel.t = 0
  // 0.6–1.4 of a nominal 1/hz period. Drawn per channel per hop, so the
  // angle and scale clocks drift apart instead of retargeting in lockstep.
  channel.dur = (0.6 + 0.8 * rand01(s)) / hz
}

function stepChannel(channel, s, dt, lo, hi, hz) {
  if (channel.dur <= 0)
    retarget(channel, s, lo, hi, hz)
  channel.t += dt
  if (channel.t >= channel.dur) {
    channel.value = channel.to
    retarget(channel, s, lo, hi, hz)
    return
  }
  var u = channel.t / channel.dur
  u = u * u * (3 - 2 * u)
  channel.value = channel.from + (channel.to - channel.from) * u
}

function step(s, dt, params) {
  var hz = params && params.hz
  if (!(hz > 0) || !(dt > 0))
    return
  var range = params.angleRangeRad
  if (!(range > 0))
    range = 0
  var lo = params.scaleMin
  var hi = params.scaleMax
  if (lo > hi) {
    var tmp = lo
    lo = hi
    hi = tmp
  }
  stepChannel(s.angle, s, dt, -range, range, hz)
  stepChannel(s.scale, s, dt, lo, hi, hz)
}

function lobe(lobeWidth, scale) {
  var v = lobeWidth * scale
  if (v < 0.04)
    return 0.04
  if (v > 0.5)
    return 0.5
  return v
}

function thickScale(scale) {
  var v = 1 + (scale - 1) * 0.35
  return v < 0.25 ? 0.25 : v
}

File: hypr/src/shimmer.cpp
#include "shimmer.hpp"

#include <algorithm>
#include <cmath>

bool shinyPulseShouldRun(bool enabled, bool pulse, float pulseHz, bool activeOnly, bool focused) {
    if (!enabled || !pulse || pulseHz <= 0.f)
        return false;
    if (activeOnly && !focused)
        return false;
    return true;
}

float shinyPulseAlphaMul(float brightness, float time) {
    if (!(brightness > 0.f))
        return 1.f;
    constexpr float kTau = 6.28318530718f;
    return 0.5f + 0.5f * std::sin(time * brightness * kTau);
}

ShinyPulseUniforms shinyPulseUniforms(bool pulse, double clockSeconds, float hz) {
    if (!pulse || hz <= 0.f)
        return {.time = 0.f, .pulseHz = 0.f};
    const double wrapped = std::fmod(clockSeconds, 1.0 / static_cast<double>(hz));
    return {.time = static_cast<float>(wrapped), .pulseHz = hz};
}

int shinyPulseTickMs(float pulseHz) {
    // ~32 samples per sine cycle, clamped so a slow pulse still breathes
    // (not one damage per 1/Hz cycle) and a fast pulse does not exceed ~60fps.
    constexpr int kMinMs = 16;
    constexpr int kMaxMs = 50;
    if (pulseHz <= 0.f)
        return kMinMs;
    const float cycleMs = 1000.f / pulseHz;
    const int   sampled = static_cast<int>(cycleMs / 32.f);
    if (sampled < kMinMs)
        return kMinMs;
    if (sampled > kMaxMs)
        return kMaxMs;
    return sampled;
}

ShinyEffect shinyEffectMode(bool pulse, float pulseHz, bool shimmer, float shimmerHz) {
    if (shimmer && shimmerHz > 0.f)
        return SHINY_EFFECT_SHIMMER;
    if (pulse && pulseHz > 0.f)
        return SHINY_EFFECT_PULSE;
    return SHINY_EFFECT_NONE;
}

bool shinyEffectShouldRun(bool enabled, ShinyEffect mode, bool activeOnly, bool focused) {
    if (!enabled || mode == SHINY_EFFECT_NONE)
        return false;
    if (activeOnly && !focused)
        return false;
    return true;
}

int shinyEffectTickMs(ShinyEffect mode, float pulseHz, float shimmerHz) {
    if (mode == SHINY_EFFECT_SHIMMER)
        return shinyPulseTickMs(shimmerHz);
    return shinyPulseTickMs(pulseHz);
}

float shinyWrapAngle(float radians) {
    const float turn = 2.f * std::acos(-1.f);
    float       r    = std::fmod(radians, turn);
    if (r < 0.f)
        r += turn;
    return r;
}

float shinyPinnedHeading(int pinDeg, int offsetDeg) {
    const float pi = std::acos(-1.f);
    return shinyWrapAngle(static_cast<float>(pinDeg + offsetDeg) * pi / 180.f);
}

static uint32_t shinyXorshift32(uint32_t& s) {
    s ^= s << 13;
    s ^= s >> 17;
    s ^= s << 5;
    return s;
}

static float shinyRand01(uint32_t& s) {
    return static_cast<float>(shinyXorshift32(s)) * (1.f / 4294967296.f);
}

void shinyShimmerSeed(ShinyShimmerState& s, uint32_t seed) {
    s.rng = seed != 0 ? seed : 0x9E3779B9u;
}

static void shinyShimmerRetarget(ShinyShimmerChannel& c, uint32_t& rng, float lo, float hi, float hz) {
    c.from = c.value;
    c.to   = lo + (hi - lo) * shinyRand01(rng);
    c.t    = 0.f;
    // 0.6–1.4 of a nominal 1/hz period. Drawn per channel per hop, so the
    // angle and scale clocks drift apart instead of retargeting in lockstep.
    c.dur  = (0.6f + 0.8f * shinyRand01(rng)) / hz;
}

static void shinyShimmerStepChannel(ShinyShimmerChannel& c, uint32_t& rng, float dt, float lo, float hi, float hz) {
    if (c.dur <= 0.f)
        shinyShimmerRetarget(c, rng, lo, hi, hz);
    c.t += dt;
    if (c.t >= c.dur) {
        c.value = c.to;
        shinyShimmerRetarget(c, rng, lo, hi, hz);
        return;
    }
    float u = c.t / c.dur;
    u       = u * u * (3.f - 2.f * u); // smoothstep ease
    c.value = c.from + (c.to - c.from) * u;
}

void shinyShimmerStep(ShinyShimmerState& s, float dt, const ShinyShimmerParams& p) {
    if (p.hz <= 0.f || dt <= 0.f)
        return;
    const float range = std::max(p.angleRangeRad, 0.f);
    const float lo    = std::min(p.scaleMin, p.scaleMax);
    const float hi    = std::max(p.scaleMin, p.scaleMax);
    shinyShimmerStepChannel(s.angle, s.rng, dt, -range, range, p.hz);
    shinyShimmerStepChannel(s.scale, s.rng, dt, lo, hi, p.hz);
}

float shinyShimmerLobe(float lobe, float scale) {
    return std::clamp(lobe * scale, 0.04f, 0.5f);
}

float shinyShimmerThickScale(float scale) {
    return std::max(1.f + (scale - 1.f) * 0.35f, 0.25f);
}
